src/xtgeo/surface/_regsurf_cube_window_v3.py

```diff
@@ -0,0 +1,373 @@
+# -*- coding: utf-8 -*-
+"""Regular surface vs Cube, slice a window interval v3, in pure numpy.
+
+The _regsurf_ variable refers to the self instance
+"""
+
+from __future__ import annotations
+
+import warnings
+from typing import Optional, Union
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+import xtgeo
+from xtgeo.common import XTGeoDialog
+
+warnings.filterwarnings(action="ignore", message="All-NaN slice encountered")
+warnings.filterwarnings(action="ignore", message="Mean of empty slice")
+warnings.filterwarnings(action="ignore", message="Degree of freedom")
+
+xtg = XTGeoDialog()
+
+logger = xtg.functionlogger(__name__)
+
+
+STATATTRS = [
+    "min",
+    "max",
+    "mean",
+    "var",
+    "rms",
+    "maxpos",
+    "maxneg",
+    "maxabs",
+    "meanabs",
+    "meanpos",
+    "meanneg",
+]
+SUMATTRS = [
+    "sumpos",
+    "sumneg",
+    "sumabs",
+]
+ALLATTRS = STATATTRS + SUMATTRS
+
+
+# Note: self -> _regsurf_
+
+
+def _cut_cube_deadtraces(cube: xtgeo.Cube, deadtraces: bool) -> np.ndarray:
+    """Take the cube numpy values and filter away dead traces as np.nan."""
+    logger.info("Assign dead traces")
+    dvalues = cube.values.copy()
+
+    if deadtraces and 2 in cube.traceidcodes:
+        dvalues[cube.traceidcodes == 2] = np.nan
+        logger.info("Dead traces encountered in this cube, set values to np.nan")
+    else:
+        logger.info("No dead traces encountered in this cube, or deadtraces is False")
+
+    return dvalues
+
+
+def _get_iso_maskthreshold_surface(
+    upper: xtgeo.RegularSurface,
+    lower: xtgeo.RegularSurface,
+    maskthreshold: float,
+) -> xtgeo.RegularSurface:
+    """Return a surface with value 0 where isochore <= threshold"""
+    logger.info("Maskthreshold based on isochore")
+    result = upper.copy()
+    result.fill()
+    diff = lower - upper
+    result.values = np.ma.where(diff.values <= maskthreshold, 0, 1)
+    return result
+
+
+def _proxy_surf_outside_cube(
+    _regsurf_: xtgeo.RegularSurface,
+    cube: xtgeo.Cube,
+) -> xtgeo.RegularSurface:
+    """Proxy for the part of input surface that is outside the cube area."""
+    logger.info("Get a proxy for part of original surface being outside the cube")
+    outside = _regsurf_.copy()
+    outside.values = 0.0
+
+    tmp_surf = xtgeo.surface_from_cube(cube, 0.0)
+    boundary = tmp_surf.get_boundary_polygons()
+    outside.set_outside(boundary, 1.0)
+    return outside  # is 1 outside the cube area, 0 within the cube
+
+
+def _upper_lower_surface(
+    _regsurf_,
+    cube: xtgeo.Cube,
+    zsurf: xtgeo.RegularSurface,
+    other: xtgeo.RegularSurface,
+    other_position: str,
+    zrange: float,
+) -> list:
+    """Return upper and lower surface, sampled to cube resolution."""
```

Comment:
Doesn't this technically return a `Tuple[xtgeo.RegularSurface, xtgeo.RegularSurface]`?